lib/ansible/module_utils/urls.py

```diff
@@ -218,9 +220,109 @@ def __init__(self, message, import_traceback, module=None):
+
+        self.socket = socket
+
+        self.sslobj = sslobj = context.wrap_bio(
+            self.incoming, self.outgoing, server_hostname=server_hostname
+        )
+
+        self.read = functools.partial(self._process, True, sslobj.read)
+        self.recv = functools.partial(self._process, True, sslobj.read)
+        self.send = functools.partial(self._process, False, sslobj.write)
+        self.sendall = functools.partial(self._process, False, sslobj.write)
+        self._decref_socketios = socket._decref_socketios
+
+        self._process(False, sslobj.do_handshake)
+
+    def close(self):
+        ...
+
+    def recv_into(self, buf, nbytes=None, flags=0):
+        if nbytes is None:
+            nbytes = len(buf)
+        return self._process(True, self.sslobj.read, nbytes, buf)
+
+    def _process_out(self):
+        buf = self.outgoing.read()
+        self.socket.sendall(buf)
+
+    def _process_in(self):
+        buf = self.socket.recv()
+        if buf:
+            self.incoming.write(buf)
+        else:
+            self.incoming.write_eof()
+
+    def _process(self, read, func, *args, **kwargs):
+        while True:
+            try:
+                return func(*args, **kwargs)
+            except ssl.SSLWantReadError as e:
+                self._process_out()
+                self._process_in()
+            except ssl.SSLWantWriteError as e:
+                self._process_in()
+                self._process_out()
+            else:
+                if read:
+                    self._process_out()
+                else:
+                    self._process_in()
+                break
+
+    def makefile(self, mode="r", buffering=None, *,
+                 encoding=None, errors=None, newline=None):
+        if buffering is None or buffering < 0:
+            buffering = io.DEFAULT_BUFFER_SIZE
+        return io.BufferedReader(
+            socket.SocketIO(self, 'r'),
+            buffering
+        )
+
+
+HTTPSHandler = None
+HTTPSConnection = None
 UnixHTTPSHandler = None
 UnixHTTPSConnection = None
 if HAS_SSL:
+    class HTTPSConnection(http.client.HTTPSConnection):  # type: ignore[no-redef]
+        def __init__(self, *args, **kwargs):
+            self._tls_in_tls = kwargs.pop('tls_in_tls', False)
+            super().__init__(*args, **kwargs)
+
+        def connect(self):
+            super().connect()
+            if self._tunnel_host:
+                super()._tunnel()
+                if self._tls_in_tls:
+                    self.sock = TLSinTLS(self.sock, self._context, self._tunnel_host)
+
+        def _tunnel(self):
+            """This is a noop, we don't want http.client.HTTPConnection._tunnel running before we have
+            a chance to have HTTPSConnection wrap the socket for HTTPS
+
+            The fact that http.client.HTTPConnection.connect establishes the tunnel is the 1st reason
+            HTTPS proxies aren't supported, the 2nd is that there is no native TLS in TLS, see TLSinTLS
+            """
```

Comment:
Overriding `_tunnel` to not do anything, and then calling `super()._tunnel()` are probably the biggest concerns with fragility and supportability here, but urllib3 does it too, kind of:  

https://github.com/urllib3/urllib3/blob/b7c2910a9f49ba0d58c8d4381500c208f9a24765/src/urllib3/connection.py#L242
https://github.com/urllib3/urllib3/blob/b7c2910a9f49ba0d58c8d4381500c208f9a24765/src/urllib3/connection.py#L631

Reply:
This doesn't look like it should be a docstring but rather a comment. Could you change?